// @remotion/media-parser/dist/readers/from-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeReader = void 0;
const fs_1 = require("fs");
const promises_1 = require("node:fs/promises");
const path_1 = require("path");
const stream_1 = require("stream");
exports.nodeReader = {
    read: async (src, range, signal) => {
        if (typeof src !== 'string') {
            throw new Error('src must be a string when using `nodeReader`');
        }
        const controller = new AbortController();
        const stream = (0, fs_1.createReadStream)(src, {
            start: range === null ? 0 : typeof range === 'number' ? range : range[0],
            end: range === null
                ? Infinity
                : typeof range === 'number'
                    ? Infinity
                    : range[1],
            signal: controller.signal,
        });
        signal === null || signal === void 0 ? void 0 : signal.addEventListener('abort', () => {
            controller.abort();
        }, { once: true });
        const stats = await (0, promises_1.stat)(src);
        const reader = stream_1.Readable.toWeb(stream).getReader();
        if (signal) {
            signal.addEventListener('abort', () => {
                reader.cancel().catch(() => { });
            }, { once: true });
        }
        return {
            reader: {
                reader,
                abort: () => {
                    controller.abort();
                },
            },
            contentLength: stats.size,
            contentType: null,
            name: src.split(path_1.sep).pop(),
            supportsContentRange: true,
        };
    },
    getLength: async (src) => {
        if (typeof src !== 'string') {
            throw new Error('src must be a string when using `nodeReader`');
        }
        const stats = await (0, promises_1.stat)(src);
        return stats.size;
    },
};
