// @remotion/media-parser/dist/boxes/riff/expect-riff-box.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expectRiffBox = void 0;
const is_movi_1 = require("./is-movi");
const parse_movi_1 = require("./parse-movi");
const parse_riff_box_1 = require("./parse-riff-box");
const expectRiffBox = async ({ iterator, state, structure, }) => {
    // Need at least 16 bytes to read LIST,size,movi,size
    if (iterator.bytesRemaining() < 16) {
        return {
            type: 'incomplete',
            continueParsing() {
                return (0, exports.expectRiffBox)({ structure, iterator, state });
            },
        };
    }
    const ckId = iterator.getByteString(4, false);
    const ckSize = iterator.getUint32Le();
    if ((0, is_movi_1.isMoviAtom)(iterator, ckId)) {
        iterator.discard(4);
        return (0, parse_movi_1.parseMovi)({
            iterator,
            maxOffset: ckSize + iterator.counter.getOffset() - 4,
            state,
            structure,
        });
    }
    if (iterator.bytesRemaining() < ckSize) {
        iterator.counter.decrement(8);
        return {
            type: 'incomplete',
            continueParsing: () => {
                return (0, exports.expectRiffBox)({ structure, iterator, state });
            },
        };
    }
    return {
        type: 'complete',
        box: await (0, parse_riff_box_1.parseRiffBox)({
            id: ckId,
            iterator,
            size: ckSize,
            boxes: structure.boxes,
            state,
        }),
        skipTo: null,
    };
};
exports.expectRiffBox = expectRiffBox;
