// @remotion/media-parser/dist/boxes/iso-base-media/stsd/stsd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseStsd = void 0;
const samples_1 = require("./samples");
const parseStsd = async ({ iterator, offset, size, state, signal, fields, }) => {
    const version = iterator.getUint8();
    if (version !== 0) {
        throw new Error(`Unsupported STSD version ${version}`);
    }
    // flags, we discard them
    iterator.discard(3);
    const numberOfEntries = iterator.getUint32();
    const bytesRemainingInBox = size - (iterator.counter.getOffset() - offset);
    const boxes = await (0, samples_1.parseSamples)({
        iterator,
        maxBytes: bytesRemainingInBox,
        state,
        signal,
        logLevel: 'info',
        fields,
    });
    if (boxes.length !== numberOfEntries) {
        throw new Error(`Expected ${numberOfEntries} sample descriptions, got ${boxes.length}`);
    }
    return {
        type: 'stsd-box',
        boxSize: size,
        offset,
        numberOfEntries,
        samples: boxes,
    };
};
exports.parseStsd = parseStsd;
