// @remotion/media-parser/dist/boxes/webm/parse-webm-header.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseWebm = void 0;
const parse_children_1 = require("./segments/parse-children");
const continueAfterMatroskaResult = (result, structure) => {
    if (result.status === 'done') {
        return {
            status: 'done',
        };
    }
    return {
        status: 'incomplete',
        continueParsing: async () => {
            const newResult = await result.continueParsing();
            return continueAfterMatroskaResult(newResult, structure);
        },
        skipTo: null,
    };
};
// Parsing according to https://darkcoding.net/software/reading-mediarecorders-webm-opus-output/
const parseWebm = async ({ counter, state, fields, }) => {
    const structure = state.structure.getStructure();
    if (structure.type !== 'matroska') {
        throw new Error('Invalid structure type');
    }
    const results = await (0, parse_children_1.expectChildren)({
        iterator: counter,
        length: Infinity,
        children: structure.boxes,
        state,
        startOffset: counter.counter.getOffset(),
        fields,
        topLevelStructure: structure,
    });
    return continueAfterMatroskaResult(results, structure);
};
exports.parseWebm = parseWebm;
