// @remotion/media-parser/dist/state/can-skip-tracks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCanSkipTracksState = void 0;
const needsTracksField = {
    audioCodec: true,
    container: false,
    dimensions: true,
    durationInSeconds: true,
    slowDurationInSeconds: true,
    slowFps: true,
    fps: true,
    internalStats: false,
    isHdr: true,
    name: false,
    rotation: true,
    size: false,
    structure: true,
    tracks: true,
    unrotatedDimensions: true,
    videoCodec: true,
    metadata: true,
    location: true,
    mimeType: false,
    slowKeyframes: true,
    slowNumberOfFrames: true,
    keyframes: true,
};
const makeCanSkipTracksState = ({ hasAudioTrackHandlers, fields, hasVideoTrackHandlers, }) => {
    return {
        canSkipTracks: () => {
            if (hasAudioTrackHandlers || hasVideoTrackHandlers) {
                return false;
            }
            const keys = Object.keys(fields !== null && fields !== void 0 ? fields : {});
            const selectedKeys = keys.filter((k) => fields[k]);
            return !selectedKeys.some((k) => needsTracksField[k]);
        },
    };
};
exports.makeCanSkipTracksState = makeCanSkipTracksState;
