// @remotion/media-parser/dist/state/parser-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeParserState = void 0;
const emitted_fields_1 = require("./emitted-fields");
const keyframes_1 = require("./keyframes");
const riff_1 = require("./riff");
const sample_callbacks_1 = require("./sample-callbacks");
const slow_duration_fps_1 = require("./slow-duration-fps");
const structure_1 = require("./structure");
const webm_1 = require("./webm");
const makeParserState = ({ hasAudioTrackHandlers, hasVideoTrackHandlers, signal, getIterator, fields, onAudioTrack, onVideoTrack, supportsContentRange, }) => {
    let skippedBytes = 0;
    const increaseSkippedBytes = (bytes) => {
        skippedBytes += bytes;
    };
    const structure = (0, structure_1.structureState)();
    const keyframes = (0, keyframes_1.keyframesState)();
    const emittedFields = (0, emitted_fields_1.emittedState)();
    const slowDurationAndFps = (0, slow_duration_fps_1.slowDurationAndFpsState)();
    return {
        riff: (0, riff_1.riffSpecificState)(),
        callbacks: (0, sample_callbacks_1.sampleCallback)({
            signal,
            hasAudioTrackHandlers,
            hasVideoTrackHandlers,
            fields,
            keyframes,
            emittedFields,
            slowDurationAndFpsState: slowDurationAndFps,
        }),
        getInternalStats: () => {
            var _a, _b;
            return ({
                skippedBytes,
                finalCursorOffset: (_b = (_a = getIterator()) === null || _a === void 0 ? void 0 : _a.counter.getOffset()) !== null && _b !== void 0 ? _b : 0,
            });
        },
        getSkipBytes: () => skippedBytes,
        increaseSkippedBytes,
        keyframes,
        structure,
        onAudioTrack,
        onVideoTrack,
        supportsContentRange,
        webm: (0, webm_1.webmState)(),
        emittedFields,
        fields,
        slowDurationAndFps,
    };
};
exports.makeParserState = makeParserState;
