// @remotion/media-parser/dist/get-dimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasDimensions = exports.getDimensions = void 0;
const get_tracks_1 = require("./get-tracks");
const getDimensions = (boxes, state) => {
    const { videoTracks } = (0, get_tracks_1.getTracks)(boxes, state);
    if (!videoTracks.length) {
        throw new Error('Expected video track');
    }
    const firstVideoTrack = videoTracks[0];
    return {
        width: firstVideoTrack.width,
        height: firstVideoTrack.height,
        rotation: firstVideoTrack.rotation,
        unrotatedHeight: firstVideoTrack.displayAspectHeight,
        unrotatedWidth: firstVideoTrack.displayAspectWidth,
    };
};
exports.getDimensions = getDimensions;
// TODO: An audio track should return 'hasDimensions' = true on an audio file
// and stop parsing
const hasDimensions = (boxes, state) => {
    try {
        return (0, exports.getDimensions)(boxes, state) !== null;
    }
    catch (_a) {
        return false;
    }
};
exports.hasDimensions = hasDimensions;
